### PR TITLE
Start documenting how to write frontend extensions

### PR DESCRIPTION
**Pull request by Carreau**
This if course require more work, but the request for custom key binding
is too high, and the process of writing extensions too widespread on the
internet.

---- 

This is the base to write an extension, @michaelpacer requested on a private communication how to also bind `^[` to go-to-command-mode / Esc, I''m writing that to see if it's sufficient information. 

@michaelpacer I'll give you push access to **my** fork, so that you can update the PR if there are issues, and fix english mistakes that will make you eyes bleed. If it's unclear I can also update the relevant part. We can hopefully reach quickly an acceptable state.

@JanSchulz , @damianavila and @juhasch also write a lot of extensions and might want to give their input on that.

### CODE REVIEW

**Comment by juhasch** [comment]
This is a good idea, IMO. Currently I cannot be much of help,  hopefully in a few weeks.

@jcb91 is also very active, maybe he has some comments/additions ?

**Comment by jcb91** [comment]
@juhasch thanks for ping-ing me in on this (not really sure what the appropriate word for the action is, or even if one exists, so ping will have to do!).

@Carreau good plan, I recall frequently coming to
[your docs on extensions](https://github.com/Carreau/jupyter-book/blob/master/notebook-extensions.md) after trying to google how to alter some notebook behaviour.
I'd be happy to contribute some time to improve/extend/update docs.
Would you prefer comments on here, or PRs to your fork?

**Comment by jcb91** [comment]
As a side note, I've been contemplating for a little while trying to write a hotkey-rebinder extension, which could be used to neatly rebind any named actions (calling javascript extensions without named actions is obviously a bit more difficult). I'll post back here if/when I get round to it.

**Comment by takluyver** [comment]
I have opened Carreau/jupyter_notebook#6 with some improvements to this.

**Comment by Carreau** [comment]
> I have opened Carreau/jupyter_notebook#6 with some improvements to this.

Merged.

File: docs/source/extending/frontend_extensions.rst [hunk 1]
```diff
@@ -0,0 +1,138 @@
+Custom front-end extesions
```
Comment (minrk) [inline docs/source/extending/frontend_extensions.rst]: extensions

Comment (Carreau) [inline docs/source/extending/frontend_extensions.rst]: Done.

**Comment by damianavila** [comment]
A little bit late, but LGTM, please also check this which is closely related: https://github.com/jupyter/enhancement-proposals/pull/7

**Comment by jcb91** [comment]
So, as far as altering keyboard shortcuts goes, I made [this nbextension](https://github.com/ipython-contrib/IPython-notebook-extensions/pull/440). It doesn't do CodeMirror shortcuts yet, as they seem more complicated, but I expect it could.

**Pull request merged**